/audit crashes if more than 10 cases

### Overview

When using `/audit` on a user who has more than 10 reported cases, the command crashes with an exception:

> Cannot have more than 10 embeds in a message!

```java
8:02:04.098 [pool-2-thread-4] ERROR net.dv8tion.jda.api.requests.RestAction - RestAction queue returned failure
java.lang.IllegalArgumentException: Cannot have more than 10 embeds in a message!
	at net.dv8tion.jda.internal.utils.Checks.check(Checks.java:48) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.WebhookMessageActionImpl.addEmbeds(WebhookMessageActionImpl.java:138) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.entities.AbstractWebhookClient.sendMessageEmbeds(AbstractWebhookClient.java:67) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.entities.AbstractWebhookClient.sendMessageEmbeds(AbstractWebhookClient.java:40) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at org.togetherjava.tjbot.commands.moderation.AuditCommand.lambda$auditUser$5(AuditCommand.java:159) ~[main/:?]
	at net.dv8tion.jda.internal.requests.restaction.operator.FlatMapRestAction.supply(FlatMapRestAction.java:44) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.FlatMapRestAction.lambda$queue$0(FlatMapRestAction.java:54) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.RestActionOperator.doSuccess(RestActionOperator.java:44) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.MapRestAction.lambda$queue$0(MapRestAction.java:41) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.RestActionOperator.doSuccess(RestActionOperator.java:44) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.MapRestAction.lambda$queue$0(MapRestAction.java:41) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.RestActionOperator.doSuccess(RestActionOperator.java:44) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.CombineRestAction.lambda$queue$6(CombineRestAction.java:136) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.api.utils.MiscUtil.locked(MiscUtil.java:146) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.CombineRestAction.lambda$queue$7(CombineRestAction.java:129) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.RestActionOperator.doSuccess(RestActionOperator.java:44) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.MapRestAction.lambda$queue$0(MapRestAction.java:41) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.DeferredRestAction.queue(DeferredRestAction.java:138) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.MapRestAction.queue(MapRestAction.java:41) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.CombineRestAction.queue(CombineRestAction.java:129) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.MapRestAction.queue(MapRestAction.java:41) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.MapRestAction.queue(MapRestAction.java:41) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.FlatMapRestAction.queue(FlatMapRestAction.java:51) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.api.requests.RestAction.queue(RestAction.java:573) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.api.requests.RestAction.queue(RestAction.java:539) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at org.togetherjava.tjbot.commands.moderation.AuditCommand.auditUser(AuditCommand.java:160) ~[main/:?]
	at org.togetherjava.tjbot.commands.moderation.AuditCommand.onSlashCommand(AuditCommand.java:126) ~[main/:?]
	at org.togetherjava.tjbot.commands.system.CommandSystem.lambda$onSlashCommand$7(CommandSystem.java:121) ~[main/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: net.dv8tion.jda.api.exceptions.ContextException
	at net.dv8tion.jda.api.exceptions.ContextException.here(ContextException.java:54) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.RestActionOperator.contextWrap(RestActionOperator.java:105) ~[JDA-4.4.0_351.jar:4.4.0_351]
	at net.dv8tion.jda.internal.requests.restaction.operator.FlatMapRestAction.queue(FlatMapRestAction.java:50) ~[JDA-4.4.0_351.jar:4.4.0_351]
	... 8 more
```

### Possible solutions

The cause is quite clear, it attempts to display all cases and then fails cause there are too many. Obviously, this case is not supposed to be triggered really often and its not thaaat big of a deal but we should find at least _a solution_.

Dyno has **pagination support** to solve this, so there is an **optional** option on the command to show page 2, page 3, ... Would probably be a clean solution to do it that way ourselves as well.

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -40,8 +40,8 @@ public final class AuditCommand extends SlashCommandAdapter {
     private static final String COMMAND_NAME = "audit";
     private static final String ACTION_VERB = "audit";
     private static final int MAX_PAGE_LENGTH = 25;
-    private static final String PREVIOUS_BUTTON = "⬅";
-    private static final String NEXT_BUTTON = "➡";
+    private static final String PREVIOUS_BUTTON_LABEL = "⬅";
+    private static final String NEXT_BUTTON_LABEL = "➡";
     private final Predicate<String> hasRequiredRole;
     private final ModerationActionsStore actionsStore;
 
@@ -196,14 +196,14 @@ public final class AuditCommand extends SlashCommandAdapter {
     private @NotNull ActionRow makeActionRow(long guildId, long targetId, long callerId,
             int pageNumber, int totalPages) {
         int previousButtonTurnPageBy = -1;
-        Button previousButton = createPageTurnButton(PREVIOUS_BUTTON, guildId, targetId, callerId,
+        Button previousButton = createPageTurnButton(PREVIOUS_BUTTON_LABEL, guildId, targetId, callerId,
                 pageNumber, previousButtonTurnPageBy);
         if (pageNumber == 1) {
             previousButton = previousButton.asDisabled();
         }
 
-        int nextButtonTurnPageBy = +1;
-        Button nextButton = createPageTurnButton(NEXT_BUTTON, guildId, targetId, callerId,
+        int nextButtonTurnPageBy = 1;
+        Button nextButton = createPageTurnButton(NEXT_BUTTON_LABEL, guildId, targetId, callerId,
                 pageNumber, nextButtonTurnPageBy);
         if (pageNumber == totalPages) {
             nextButton = nextButton.asDisabled();

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -57,7 +57,7 @@ public final class AuditCommand extends SlashCommandAdapter {
     }
 
     private static @NotNull EmbedBuilder createSummaryEmbed(@NotNull User user,
-                                                            @NotNull Collection<ActionRecord> actions) {
+            @NotNull Collection<ActionRecord> actions) {
         return new EmbedBuilder().setTitle("Audit log of **%s**".formatted(user.getAsTag()))
             .setAuthor(user.getName(), null, user.getAvatarUrl())
             .setDescription(createSummaryMessageDescription(actions))
@@ -68,7 +68,8 @@ public final class AuditCommand extends SlashCommandAdapter {
             @NotNull Collection<ActionRecord> actions) {
         int actionAmount = actions.size();
 
-        String shortSummary = "There are **%s actions** against the user.".formatted(actionAmount == 0 ? "no" : actionAmount);
+        String shortSummary = "There are **%s actions** against the user."
+            .formatted(actionAmount == 0 ? "no" : actionAmount);
 
         // Summary of all actions with their count, like "- Warn: 5", descending
         Map<ModerationAction, Long> actionTypeToCount = actions.stream()
@@ -86,18 +87,22 @@ public final class AuditCommand extends SlashCommandAdapter {
     }
 
     private static @NotNull MessageEmbed.Field actionToField(@NotNull ActionRecord action,
-                                                               @NotNull JDA jda) {
-        Function<Instant, String> formatTime = instant -> TimeUtil.getDateTimeString(instant.atOffset(ZoneOffset.UTC));
+            @NotNull JDA jda) {
+        Function<Instant, String> formatTime = instant -> {
+            if (instant == null)
+                return "";
+            return TimeUtil.getDateTimeString(instant.atOffset(ZoneOffset.UTC));
+        };
 
-        String expiresAt = action.actionExpiresAt() == null ? "" : "\nTemporary action, Expires at: " + formatTime.apply(action.actionExpiresAt());
-
-        User targetUser = jda.getUserById(action.authorId());
+        User author = jda.getUserById(action.authorId());
 
         return new MessageEmbed.Field(
-                action.actionType().name() + " by " + (targetUser == null ? "(unknown user)" : targetUser.getAsTag()),
-                action.reason() + "\nIssued at: " + formatTime.apply(action.issuedAt()) + expiresAt,
-                false
-        );
+                action.actionType().name() + " by "
+                        + (author == null ? "(unknown user)" : author.getAsTag()),
+                action.reason() + "\nIssued at: " + formatTime.apply(action.issuedAt())
+                        + "\nTemporary action, Expires at: "
+                        + formatTime.apply(action.actionExpiresAt()),
+                false);
     }
 
     @Override

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -75,6 +75,10 @@ public final class AuditCommand extends SlashCommandAdapter {
         String shortSummary = "There are **%s actions** against the user."
             .formatted(actionAmount == 0 ? "no" : actionAmount);
 
+        if (actionAmount == 0) {
+            return shortSummary;
+        }
+
         // Summary of all actions with their count, like "- Warn: 5", descending
         Map<ModerationAction, Long> actionTypeToCount = actions.stream()
             .collect(Collectors.groupingBy(ActionRecord::actionType, Collectors.counting()));

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -159,7 +159,8 @@ public final class AuditCommand extends SlashCommandAdapter {
         return groupedActions;
     }
 
-    private @NotNull Message auditUser(long guildId, long targetId, int pageNumber, @NotNull JDA jda) {
+    private @NotNull Message auditUser(long guildId, long targetId, int pageNumber,
+            @NotNull JDA jda) {
         List<ActionRecord> actions = actionsStore.getActionsByTargetAscending(guildId, targetId);
         List<List<ActionRecord>> groupedActions = groupActions(actions);
         int totalPages = groupedActions.size();
@@ -178,17 +179,20 @@ public final class AuditCommand extends SlashCommandAdapter {
             .build();
     }
 
-    private @NotNull ActionRow makeActionRow(long guildId, long targetId, int pageNumber, int totalPages) {
+    private @NotNull ActionRow makeActionRow(long guildId, long targetId, int pageNumber,
+            int totalPages) {
         String stringGuildId = String.valueOf(guildId);
         String stringTargetId = String.valueOf(targetId);
         String stringPageNumber = String.valueOf(pageNumber);
 
-        Button previousButton = Button.primary(generateComponentId(stringGuildId, stringTargetId, stringPageNumber, "-1"), "⬅");
+        Button previousButton = Button.primary(
+                generateComponentId(stringGuildId, stringTargetId, stringPageNumber, "-1"), "⬅");
         if (pageNumber == 1) {
             previousButton = previousButton.asDisabled();
         }
 
-        Button nextButton = Button.primary(generateComponentId(stringGuildId, stringTargetId, stringPageNumber, "1"), "➡");
+        Button nextButton = Button.primary(
+                generateComponentId(stringGuildId, stringTargetId, stringPageNumber, "1"), "➡");
         if (pageNumber == totalPages) {
             nextButton = nextButton.asDisabled();
         }

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -101,12 +101,15 @@ public final class AuditCommand extends SlashCommandAdapter {
 
         User author = jda.getUserById(action.authorId());
 
+        Instant expiresAt = action.actionExpiresAt();
+        String expiresAtFormatted = expiresAt == null ? ""
+                : "\nTemporary action, expires at: " + formatTime.apply(expiresAt);
+
         return new MessageEmbed.Field(
                 action.actionType().name() + " by "
                         + (author == null ? "(unknown user)" : author.getAsTag()),
                 action.reason() + "\nIssued at: " + formatTime.apply(action.issuedAt())
-                        + "\nTemporary action, Expires at: "
-                        + formatTime.apply(action.actionExpiresAt()),
+                        + expiresAtFormatted,
                 false);
     }
 

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -127,7 +127,7 @@ public final class AuditCommand extends SlashCommandAdapter {
             return;
         }
 
-        event.reply(auditMessage(guild.getIdLong(), target.getIdLong(), 1, event.getJDA())).queue();
+        event.reply(auditUser(guild.getIdLong(), target.getIdLong(), 1, event.getJDA())).queue();
     }
 
     @SuppressWarnings("BooleanMethodNameMustStartWithQuestion")
@@ -141,7 +141,7 @@ public final class AuditCommand extends SlashCommandAdapter {
         return ModerationUtils.handleHasAuthorRole(ACTION_VERB, hasRequiredRole, author, event);
     }
 
-    private Message auditMessage(long guildId, long targetId, int pageNo, @NotNull JDA jda) {
+    private Message auditUser(long guildId, long targetId, int pageNo, @NotNull JDA jda) {
         List<ActionRecord> actions = actionsStore.getActionsByTargetAscending(guildId, targetId);
 
         List<List<ActionRecord>> groupedActions = new ArrayList<>();
@@ -195,7 +195,7 @@ public final class AuditCommand extends SlashCommandAdapter {
     @Override
     public void onButtonClick(@NotNull ButtonInteractionEvent event, @NotNull List<String> args) {
         event
-            .editMessage(auditMessage(Long.parseLong(args.get(0)), Long.parseLong(args.get(1)),
+            .editMessage(auditUser(Long.parseLong(args.get(0)), Long.parseLong(args.get(1)),
                     Integer.parseInt(args.get(2)) + Integer.parseInt(args.get(3)), event.getJDA()))
             .queue();
     }

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -39,6 +39,7 @@ public final class AuditCommand extends SlashCommandAdapter {
     private static final String TARGET_OPTION = "user";
     private static final String COMMAND_NAME = "audit";
     private static final String ACTION_VERB = "audit";
+    private static final int MAX_PAGE_LENGTH = 25;
     private final Predicate<String> hasRequiredRole;
     private final ModerationActionsStore actionsStore;
 
@@ -145,19 +146,23 @@ public final class AuditCommand extends SlashCommandAdapter {
         return ModerationUtils.handleHasAuthorRole(ACTION_VERB, hasRequiredRole, author, event);
     }
 
-    private Message auditUser(long guildId, long targetId, int pageNo, @NotNull JDA jda) {
-        List<ActionRecord> actions = actionsStore.getActionsByTargetAscending(guildId, targetId);
-
+    private @NotNull List<List<ActionRecord>> groupActions(@NotNull List<ActionRecord> actions) {
         List<List<ActionRecord>> groupedActions = new ArrayList<>();
         for (int i = 0; i < actions.size(); i++) {
-            if (i % 25 == 0) {
-                groupedActions.add(new ArrayList<>(25));
+            if (i % AuditCommand.MAX_PAGE_LENGTH == 0) {
+                groupedActions.add(new ArrayList<>(AuditCommand.MAX_PAGE_LENGTH));
             }
 
             groupedActions.get(groupedActions.size() - 1).add(actions.get(i));
         }
 
-        int totalPage = groupedActions.size();
+        return groupedActions;
+    }
+
+    private @NotNull Message auditUser(long guildId, long targetId, int pageNumber, @NotNull JDA jda) {
+        List<ActionRecord> actions = actionsStore.getActionsByTargetAscending(guildId, targetId);
+        List<List<ActionRecord>> groupedActions = groupActions(actions);
+        int totalPages = groupedActions.size();
 
         EmbedBuilder audit = createSummaryEmbed(jda.retrieveUserById(targetId).complete(), actions);
 
@@ -165,42 +170,38 @@ public final class AuditCommand extends SlashCommandAdapter {
             return new MessageBuilder(audit.build()).build();
         }
 
-        groupedActions.get(pageNo - 1)
+        groupedActions.get(pageNumber - 1)
             .forEach(action -> audit.addField(actionToField(action, jda)));
 
-        return new MessageBuilder(audit.setFooter("Page: " + pageNo + "/" + totalPage).build())
-            .setActionRows(ActionRow.of(previousButton(guildId, targetId, pageNo),
-                    nextButton(guildId, targetId, pageNo, totalPage)))
+        return new MessageBuilder(audit.setFooter("Page: " + pageNumber + "/" + totalPages).build())
+            .setActionRows(makeActionRow(guildId, targetId, pageNumber, totalPages))
             .build();
     }
 
-    private Button previousButton(long guildId, long targetId, int pageNo) {
-        Button previousButton = Button.primary(generateComponentId(String.valueOf(guildId),
-                String.valueOf(targetId), String.valueOf(pageNo), "-1"), "⬅");
+    private @NotNull ActionRow makeActionRow(long guildId, long targetId, int pageNumber, int totalPages) {
+        String stringGuildId = String.valueOf(guildId);
+        String stringTargetId = String.valueOf(targetId);
+        String stringPageNumber = String.valueOf(pageNumber);
 
-        if (pageNo == 1) {
+        Button previousButton = Button.primary(generateComponentId(stringGuildId, stringTargetId, stringPageNumber, "-1"), "⬅");
+        if (pageNumber == 1) {
             previousButton = previousButton.asDisabled();
         }
 
-        return previousButton;
-    }
-
-    private Button nextButton(long guildId, long targetId, int pageNo, int totalPage) {
-        Button nextButton = Button.primary(generateComponentId(String.valueOf(guildId),
-                String.valueOf(targetId), String.valueOf(pageNo), "1"), "➡");
-
-        if (pageNo == totalPage) {
+        Button nextButton = Button.primary(generateComponentId(stringGuildId, stringTargetId, stringPageNumber, "1"), "➡");
+        if (pageNumber == totalPages) {
             nextButton = nextButton.asDisabled();
         }
 
-        return nextButton;
+        return ActionRow.of(previousButton, nextButton);
     }
 
     @Override
     public void onButtonClick(@NotNull ButtonInteractionEvent event, @NotNull List<String> args) {
-        event
-            .editMessage(auditUser(Long.parseLong(args.get(0)), Long.parseLong(args.get(1)),
-                    Integer.parseInt(args.get(2)) + Integer.parseInt(args.get(3)), event.getJDA()))
-            .queue();
+        long guildId = Long.parseLong(args.get(0));
+        long targetId = Long.parseLong(args.get(1));
+        int pageNumber = Integer.parseInt(args.get(2)) + Integer.parseInt(args.get(3));
+
+        event.editMessage(auditUser(guildId, targetId, pageNumber, event.getJDA())).queue();
     }
 }

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -45,7 +45,7 @@ public final class AuditCommand extends SlashCommandAdapter {
 
     /**
      * Constructs an instance.
-     *
+     * 
      * @param actionsStore used to store actions issued by this command
      * @param config the config to use for this
      */
@@ -168,6 +168,10 @@ public final class AuditCommand extends SlashCommandAdapter {
         List<List<ActionRecord>> groupedActions = groupActions(actions);
         int totalPages = groupedActions.size();
 
+        // Handles the case of too low page number and too high page number
+        pageNumber = Math.max(1, pageNumber);
+        pageNumber = Math.min(totalPages, pageNumber);
+
         EmbedBuilder audit = createSummaryEmbed(jda.retrieveUserById(targetId).complete(), actions);
 
         if (groupedActions.isEmpty()) {
@@ -189,14 +193,16 @@ public final class AuditCommand extends SlashCommandAdapter {
         String stringCallerId = String.valueOf(callerId);
         String stringPageNumber = String.valueOf(pageNumber);
 
+        String previousButtonTurnPageBy = "-1";
         Button previousButton = Button.primary(generateComponentId(stringGuildId, stringTargetId,
-                stringCallerId, stringPageNumber, "-1"), "⬅");
+                stringCallerId, stringPageNumber, previousButtonTurnPageBy), "⬅");
         if (pageNumber == 1) {
             previousButton = previousButton.asDisabled();
         }
 
+        String nextButtonTurnPageBy = "1";
         Button nextButton = Button.primary(generateComponentId(stringGuildId, stringTargetId,
-                stringCallerId, stringPageNumber, "1"), "➡");
+                stringCallerId, stringPageNumber, nextButtonTurnPageBy), "➡");
         if (pageNumber == totalPages) {
             nextButton = nextButton.asDisabled();
         }
@@ -210,12 +216,16 @@ public final class AuditCommand extends SlashCommandAdapter {
         long interactorId = event.getMember().getIdLong();
 
         if (callerId == interactorId) {
+            int currentPage = Integer.parseInt(args.get(3));
+            int turnPageBy = Integer.parseInt(args.get(4));
+
             long guildId = Long.parseLong(args.get(0));
             long targetId = Long.parseLong(args.get(1));
-            int pageNumber = Integer.parseInt(args.get(3)) + Integer.parseInt(args.get(4));
+            int pageToDisplay = currentPage + turnPageBy;
 
             event
-                .editMessage(auditUser(guildId, targetId, interactorId, pageNumber, event.getJDA()))
+                .editMessage(
+                        auditUser(guildId, targetId, interactorId, pageToDisplay, event.getJDA()))
                 .queue();
         } else {
             event.reply("Only the user who triggered the command can use these buttons.")

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -89,8 +89,9 @@ public final class AuditCommand extends SlashCommandAdapter {
     private static @NotNull MessageEmbed.Field actionToField(@NotNull ActionRecord action,
             @NotNull JDA jda) {
         Function<Instant, String> formatTime = instant -> {
-            if (instant == null)
+            if (instant == null) {
                 return "";
+            }
             return TimeUtil.getDateTimeString(instant.atOffset(ZoneOffset.UTC));
         };
 

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -154,7 +154,7 @@ public final class AuditCommand extends SlashCommandAdapter {
 
         EmbedBuilder audit = createSummaryEmbed(jda.retrieveUserById(targetId).complete(), actions);
 
-        if (groupedActions.size() == 0) {
+        if (groupedActions.isEmpty()) {
             return new MessageBuilder(audit.build()).build();
         }
 

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -223,6 +223,8 @@ public final class AuditCommand extends SlashCommandAdapter {
             event.reply("Only the user who triggered the command can use these buttons.")
                 .setEphemeral(true)
                 .queue();
+
+            return;
         }
 
         int currentPage = Integer.parseInt(args.get(3));

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -132,7 +132,10 @@ public final class AuditCommand extends SlashCommandAdapter {
             return;
         }
 
-        event.reply(auditUser(guild.getIdLong(), target.getIdLong(), 1, event.getJDA())).queue();
+        event
+            .reply(auditUser(guild.getIdLong(), target.getIdLong(), event.getMember().getIdLong(),
+                    1, event.getJDA()))
+            .queue();
     }
 
     @SuppressWarnings("BooleanMethodNameMustStartWithQuestion")
@@ -159,7 +162,7 @@ public final class AuditCommand extends SlashCommandAdapter {
         return groupedActions;
     }
 
-    private @NotNull Message auditUser(long guildId, long targetId, int pageNumber,
+    private @NotNull Message auditUser(long guildId, long targetId, long callerId, int pageNumber,
             @NotNull JDA jda) {
         List<ActionRecord> actions = actionsStore.getActionsByTargetAscending(guildId, targetId);
         List<List<ActionRecord>> groupedActions = groupActions(actions);
@@ -175,24 +178,25 @@ public final class AuditCommand extends SlashCommandAdapter {
             .forEach(action -> audit.addField(actionToField(action, jda)));
 
         return new MessageBuilder(audit.setFooter("Page: " + pageNumber + "/" + totalPages).build())
-            .setActionRows(makeActionRow(guildId, targetId, pageNumber, totalPages))
+            .setActionRows(makeActionRow(guildId, targetId, callerId, pageNumber, totalPages))
             .build();
     }
 
-    private @NotNull ActionRow makeActionRow(long guildId, long targetId, int pageNumber,
-            int totalPages) {
+    private @NotNull ActionRow makeActionRow(long guildId, long targetId, long callerId,
+            int pageNumber, int totalPages) {
         String stringGuildId = String.valueOf(guildId);
         String stringTargetId = String.valueOf(targetId);
+        String stringCallerId = String.valueOf(callerId);
         String stringPageNumber = String.valueOf(pageNumber);
 
-        Button previousButton = Button.primary(
-                generateComponentId(stringGuildId, stringTargetId, stringPageNumber, "-1"), "⬅");
+        Button previousButton = Button.primary(generateComponentId(stringGuildId, stringTargetId,
+                stringCallerId, stringPageNumber, "-1"), "⬅");
         if (pageNumber == 1) {
             previousButton = previousButton.asDisabled();
         }
 
-        Button nextButton = Button.primary(
-                generateComponentId(stringGuildId, stringTargetId, stringPageNumber, "1"), "➡");
+        Button nextButton = Button.primary(generateComponentId(stringGuildId, stringTargetId,
+                stringCallerId, stringPageNumber, "1"), "➡");
         if (pageNumber == totalPages) {
             nextButton = nextButton.asDisabled();
         }
@@ -202,10 +206,21 @@ public final class AuditCommand extends SlashCommandAdapter {
 
     @Override
     public void onButtonClick(@NotNull ButtonInteractionEvent event, @NotNull List<String> args) {
-        long guildId = Long.parseLong(args.get(0));
-        long targetId = Long.parseLong(args.get(1));
-        int pageNumber = Integer.parseInt(args.get(2)) + Integer.parseInt(args.get(3));
+        long callerId = Long.parseLong(args.get(2));
+        long interactorId = event.getMember().getIdLong();
 
-        event.editMessage(auditUser(guildId, targetId, pageNumber, event.getJDA())).queue();
+        if (callerId == interactorId) {
+            long guildId = Long.parseLong(args.get(0));
+            long targetId = Long.parseLong(args.get(1));
+            int pageNumber = Integer.parseInt(args.get(3)) + Integer.parseInt(args.get(4));
+
+            event
+                .editMessage(auditUser(guildId, targetId, interactorId, pageNumber, event.getJDA()))
+                .queue();
+        } else {
+            event.reply("Only the user who triggered the command can use these buttons.")
+                .setEphemeral(true)
+                .queue();
+        }
     }
 }

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -149,7 +149,8 @@ public final class AuditCommand extends SlashCommandAdapter {
         return ModerationUtils.handleHasAuthorRole(ACTION_VERB, hasRequiredRole, author, event);
     }
 
-    private @NotNull List<List<ActionRecord>> groupActions(@NotNull List<ActionRecord> actions) {
+    private @NotNull List<List<ActionRecord>> groupActionsByPages(
+            @NotNull List<ActionRecord> actions) {
         List<List<ActionRecord>> groupedActions = new ArrayList<>();
         for (int i = 0; i < actions.size(); i++) {
             if (i % AuditCommand.MAX_PAGE_LENGTH == 0) {
@@ -162,10 +163,14 @@ public final class AuditCommand extends SlashCommandAdapter {
         return groupedActions;
     }
 
+    /**
+     * @param pageNumber page number to display when actions are divided into pages and each page
+     *        can contain {@link AuditCommand#MAX_PAGE_LENGTH} actions
+     */
     private @NotNull Message auditUser(long guildId, long targetId, long callerId, int pageNumber,
             @NotNull JDA jda) {
         List<ActionRecord> actions = actionsStore.getActionsByTargetAscending(guildId, targetId);
-        List<List<ActionRecord>> groupedActions = groupActions(actions);
+        List<List<ActionRecord>> groupedActions = groupActionsByPages(actions);
         int totalPages = groupedActions.size();
 
         // Handles the case of too low page number and too high page number
@@ -188,21 +193,20 @@ public final class AuditCommand extends SlashCommandAdapter {
 
     private @NotNull ActionRow makeActionRow(long guildId, long targetId, long callerId,
             int pageNumber, int totalPages) {
-        String stringGuildId = String.valueOf(guildId);
-        String stringTargetId = String.valueOf(targetId);
-        String stringCallerId = String.valueOf(callerId);
-        String stringPageNumber = String.valueOf(pageNumber);
-
         String previousButtonTurnPageBy = "-1";
-        Button previousButton = Button.primary(generateComponentId(stringGuildId, stringTargetId,
-                stringCallerId, stringPageNumber, previousButtonTurnPageBy), "⬅");
+        Button previousButton = Button
+            .primary(generateComponentId(String.valueOf(guildId), String.valueOf(targetId),
+                    String.valueOf(callerId), String.valueOf(pageNumber), previousButtonTurnPageBy),
+                    "⬅");
         if (pageNumber == 1) {
             previousButton = previousButton.asDisabled();
         }
 
         String nextButtonTurnPageBy = "1";
-        Button nextButton = Button.primary(generateComponentId(stringGuildId, stringTargetId,
-                stringCallerId, stringPageNumber, nextButtonTurnPageBy), "➡");
+        Button nextButton = Button.primary(
+                generateComponentId(String.valueOf(guildId), String.valueOf(targetId),
+                        String.valueOf(callerId), String.valueOf(pageNumber), nextButtonTurnPageBy),
+                "➡");
         if (pageNumber == totalPages) {
             nextButton = nextButton.asDisabled();
         }
@@ -215,22 +219,20 @@ public final class AuditCommand extends SlashCommandAdapter {
         long callerId = Long.parseLong(args.get(2));
         long interactorId = event.getMember().getIdLong();
 
-        if (callerId == interactorId) {
-            int currentPage = Integer.parseInt(args.get(3));
-            int turnPageBy = Integer.parseInt(args.get(4));
-
-            long guildId = Long.parseLong(args.get(0));
-            long targetId = Long.parseLong(args.get(1));
-            int pageToDisplay = currentPage + turnPageBy;
-
-            event
-                .editMessage(
-                        auditUser(guildId, targetId, interactorId, pageToDisplay, event.getJDA()))
-                .queue();
-        } else {
+        if (callerId != interactorId) {
             event.reply("Only the user who triggered the command can use these buttons.")
                 .setEphemeral(true)
                 .queue();
         }
+
+        int currentPage = Integer.parseInt(args.get(3));
+        int turnPageBy = Integer.parseInt(args.get(4));
+
+        long guildId = Long.parseLong(args.get(0));
+        long targetId = Long.parseLong(args.get(1));
+        int pageToDisplay = currentPage + turnPageBy;
+
+        event.editMessage(auditUser(guildId, targetId, interactorId, pageToDisplay, event.getJDA()))
+            .queue();
     }
 }

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -196,8 +196,8 @@ public final class AuditCommand extends SlashCommandAdapter {
     private @NotNull ActionRow makeActionRow(long guildId, long targetId, long callerId,
             int pageNumber, int totalPages) {
         int previousButtonTurnPageBy = -1;
-        Button previousButton = createPageTurnButton(PREVIOUS_BUTTON_LABEL, guildId, targetId, callerId,
-                pageNumber, previousButtonTurnPageBy);
+        Button previousButton = createPageTurnButton(PREVIOUS_BUTTON_LABEL, guildId, targetId,
+                callerId, pageNumber, previousButtonTurnPageBy);
         if (pageNumber == 1) {
             previousButton = previousButton.asDisabled();
         }

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -40,6 +40,8 @@ public final class AuditCommand extends SlashCommandAdapter {
     private static final String COMMAND_NAME = "audit";
     private static final String ACTION_VERB = "audit";
     private static final int MAX_PAGE_LENGTH = 25;
+    private static final String PREVIOUS_BUTTON = "⬅";
+    private static final String NEXT_BUTTON = "➡";
     private final Predicate<String> hasRequiredRole;
     private final ModerationActionsStore actionsStore;
 
@@ -193,25 +195,28 @@ public final class AuditCommand extends SlashCommandAdapter {
 
     private @NotNull ActionRow makeActionRow(long guildId, long targetId, long callerId,
             int pageNumber, int totalPages) {
-        String previousButtonTurnPageBy = "-1";
-        Button previousButton = Button
-            .primary(generateComponentId(String.valueOf(guildId), String.valueOf(targetId),
-                    String.valueOf(callerId), String.valueOf(pageNumber), previousButtonTurnPageBy),
-                    "⬅");
+        int previousButtonTurnPageBy = -1;
+        Button previousButton = createPageTurnButton(PREVIOUS_BUTTON, guildId, targetId, callerId,
+                pageNumber, previousButtonTurnPageBy);
         if (pageNumber == 1) {
             previousButton = previousButton.asDisabled();
         }
 
-        String nextButtonTurnPageBy = "1";
-        Button nextButton = Button.primary(
-                generateComponentId(String.valueOf(guildId), String.valueOf(targetId),
-                        String.valueOf(callerId), String.valueOf(pageNumber), nextButtonTurnPageBy),
-                "➡");
+        int nextButtonTurnPageBy = +1;
+        Button nextButton = createPageTurnButton(NEXT_BUTTON, guildId, targetId, callerId,
+                pageNumber, nextButtonTurnPageBy);
         if (pageNumber == totalPages) {
             nextButton = nextButton.asDisabled();
         }
 
         return ActionRow.of(previousButton, nextButton);
+    }
+
+    private @NotNull Button createPageTurnButton(@NotNull String label, long guildId, long targetId,
+            long callerId, long pageNumber, int turnPageBy) {
+        return Button.primary(generateComponentId(String.valueOf(guildId), String.valueOf(targetId),
+                String.valueOf(callerId), String.valueOf(pageNumber), String.valueOf(turnPageBy)),
+                label);
     }
 
     @Override

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -1,6 +1,5 @@
 package org.togetherjava.tjbot.commands.moderation;
 
-import io.r2dbc.spi.Parameter;
 import net.dv8tion.jda.api.EmbedBuilder;
 import net.dv8tion.jda.api.JDA;
 import net.dv8tion.jda.api.MessageBuilder;

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -1,12 +1,17 @@
 package org.togetherjava.tjbot.commands.moderation;
 
+import io.r2dbc.spi.Parameter;
 import net.dv8tion.jda.api.EmbedBuilder;
 import net.dv8tion.jda.api.JDA;
+import net.dv8tion.jda.api.MessageBuilder;
 import net.dv8tion.jda.api.entities.*;
 import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
+import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
 import net.dv8tion.jda.api.interactions.callbacks.IReplyCallback;
 import net.dv8tion.jda.api.interactions.commands.OptionMapping;
 import net.dv8tion.jda.api.interactions.commands.OptionType;
+import net.dv8tion.jda.api.interactions.components.ActionRow;
+import net.dv8tion.jda.api.interactions.components.buttons.Button;
 import net.dv8tion.jda.api.utils.TimeUtil;
 
 import org.jetbrains.annotations.NotNull;
@@ -120,7 +125,12 @@ public final class AuditCommand extends SlashCommandAdapter {
             return;
         }
 
-        auditUser(target, guild, event);
+        event.reply(auditMessage(
+                guild.getIdLong(),
+                target.getIdLong(),
+                1,
+                event.getJDA()
+        )).queue();
     }
 
     @SuppressWarnings("BooleanMethodNameMustStartWithQuestion")
@@ -134,14 +144,69 @@ public final class AuditCommand extends SlashCommandAdapter {
         return ModerationUtils.handleHasAuthorRole(ACTION_VERB, hasRequiredRole, author, event);
     }
 
-    private void auditUser(@NotNull User user, @NotNull ISnowflake guild,
-            @NotNull IReplyCallback event) {
-        List<ActionRecord> actions =
-                actionsStore.getActionsByTargetAscending(guild.getIdLong(), user.getIdLong());
+    private Message auditMessage(long guildId, long targetId, int pageNo, @NotNull JDA jda) {
+        List<ActionRecord> actions = actionsStore.getActionsByTargetAscending(guildId, targetId);
 
-        EmbedBuilder audit = createSummaryEmbed(user, actions);
-        actions.forEach(action -> audit.addField(actionToField(action, event.getJDA())));
+        List<List<ActionRecord>> groupedActions = new ArrayList<>();
+        for (int i = 0; i < actions.size(); i++) {
+            if (i % 25 == 0) {
+                groupedActions.add(new ArrayList<>(25));
+            }
 
-        event.replyEmbeds(audit.build()).queue();
+            groupedActions.get(groupedActions.size() - 1).add(actions.get(i));
+        }
+
+        int totalPage = groupedActions.size();
+
+        EmbedBuilder audit = createSummaryEmbed(jda.retrieveUserById(targetId).complete(), actions)
+                .setFooter("Page: " + pageNo + "/" + totalPage);
+        groupedActions.get(pageNo - 1).forEach(action -> audit.addField(actionToField(action, jda)));
+
+        return new MessageBuilder(audit.build())
+                .setActionRows(ActionRow.of(
+                        previousButton(guildId, targetId, pageNo),
+                        nextButton(guildId, targetId, pageNo, totalPage)
+                ))
+                .build();
+    }
+
+    private Button previousButton(long guildId, long targetId, int pageNo) {
+        Button previousButton = Button.primary(generateComponentId(
+                String.valueOf(guildId),
+                String.valueOf(targetId),
+                String.valueOf(pageNo),
+                "-1"
+        ), "⬅");
+
+        if (pageNo == 1) {
+            previousButton = previousButton.asDisabled();
+        }
+
+        return previousButton;
+    }
+
+    private Button nextButton(long guildId, long targetId, int pageNo, int totalPage) {
+        Button nextButton = Button.primary(generateComponentId(
+                String.valueOf(guildId),
+                String.valueOf(targetId),
+                String.valueOf(pageNo),
+                "1"
+        ), "➡");
+
+        if (pageNo == totalPage) {
+            nextButton = nextButton.asDisabled();
+        }
+
+        return nextButton;
+    }
+
+    @Override
+    public void onButtonClick(@NotNull ButtonInteractionEvent event, @NotNull List<String> args) {
+        event.editMessage(auditMessage(
+                Long.parseLong(args.get(0)),
+                Long.parseLong(args.get(1)),
+                Integer.parseInt(args.get(2)) + Integer.parseInt(args.get(3)),
+                event.getJDA()
+        )).queue();
     }
 }

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -7,16 +7,19 @@ import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEve
 import net.dv8tion.jda.api.interactions.callbacks.IReplyCallback;
 import net.dv8tion.jda.api.interactions.commands.OptionMapping;
 import net.dv8tion.jda.api.interactions.commands.OptionType;
-import net.dv8tion.jda.api.requests.RestAction;
 import net.dv8tion.jda.api.utils.TimeUtil;
+
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+
 import org.togetherjava.tjbot.commands.SlashCommandAdapter;
 import org.togetherjava.tjbot.commands.SlashCommandVisibility;
 import org.togetherjava.tjbot.config.Config;
 
+import java.time.Instant;
 import java.time.ZoneOffset;
 import java.util.*;
+import java.util.function.Function;
 import java.util.function.Predicate;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -53,27 +56,24 @@ public final class AuditCommand extends SlashCommandAdapter {
         this.actionsStore = Objects.requireNonNull(actionsStore);
     }
 
-    private static @NotNull MessageEmbed createSummaryMessage(@NotNull User user,
-            @NotNull Collection<ActionRecord> actions) {
+    private static @NotNull EmbedBuilder createSummaryEmbed(@NotNull User user,
+                                                            @NotNull Collection<ActionRecord> actions) {
         return new EmbedBuilder().setTitle("Audit log of **%s**".formatted(user.getAsTag()))
             .setAuthor(user.getName(), null, user.getAvatarUrl())
             .setDescription(createSummaryMessageDescription(actions))
-            .setColor(ModerationUtils.AMBIENT_COLOR)
-            .build();
+            .setColor(ModerationUtils.AMBIENT_COLOR);
     }
 
     private static @NotNull String createSummaryMessageDescription(
             @NotNull Collection<ActionRecord> actions) {
         int actionAmount = actions.size();
-        if (actionAmount == 0) {
-            return "There are **no actions** against the user.";
-        }
 
-        String shortSummary = "There are **%d actions** against the user.".formatted(actionAmount);
+        String shortSummary = "There are **%s actions** against the user.".formatted(actionAmount == 0 ? "no" : actionAmount);
 
         // Summary of all actions with their count, like "- Warn: 5", descending
         Map<ModerationAction, Long> actionTypeToCount = actions.stream()
             .collect(Collectors.groupingBy(ActionRecord::actionType, Collectors.counting()));
+
         String typeCountSummary = actionTypeToCount.entrySet()
             .stream()
             .filter(typeAndCount -> typeAndCount.getValue() > 0)
@@ -85,30 +85,19 @@ public final class AuditCommand extends SlashCommandAdapter {
         return shortSummary + "\n" + typeCountSummary;
     }
 
-    private static @NotNull RestAction<MessageEmbed> actionToMessage(@NotNull ActionRecord action,
-            @NotNull JDA jda) {
-        String footer = action.actionExpiresAt() == null ? null
-                : "Temporary action, expires at %s".formatted(TimeUtil
-                    .getDateTimeString(action.actionExpiresAt().atOffset(ZoneOffset.UTC)));
+    private static @NotNull MessageEmbed.Field actionToField(@NotNull ActionRecord action,
+                                                               @NotNull JDA jda) {
+        Function<Instant, String> formatTime = instant -> TimeUtil.getDateTimeString(instant.atOffset(ZoneOffset.UTC));
 
-        return jda.retrieveUserById(action.authorId())
-            .onErrorMap(error -> null)
-            .map(author -> new EmbedBuilder().setTitle(action.actionType().name())
-                .setAuthor(author == null ? "(unknown user)" : author.getAsTag(), null,
-                        author == null ? null : author.getAvatarUrl())
-                .setDescription(action.reason())
-                .setTimestamp(action.issuedAt())
-                .setFooter(footer)
-                .setColor(ModerationUtils.AMBIENT_COLOR)
-                .build());
-    }
+        String expiresAt = action.actionExpiresAt() == null ? "" : "\nTemporary action, Expires at: " + formatTime.apply(action.actionExpiresAt());
 
-    private static <E> @NotNull List<E> prependElement(@NotNull E element,
-            @NotNull Collection<? extends E> elements) {
-        List<E> allElements = new ArrayList<>(elements.size() + 1);
-        allElements.add(element);
-        allElements.addAll(elements);
-        return allElements;
+        User targetUser = jda.getUserById(action.authorId());
+
+        return new MessageEmbed.Field(
+                action.actionType().name() + " by " + (targetUser == null ? "(unknown user)" : targetUser.getAsTag()),
+                action.reason() + "\nIssued at: " + formatTime.apply(action.issuedAt()) + expiresAt,
+                false
+        );
     }
 
     @Override
@@ -121,7 +110,7 @@ public final class AuditCommand extends SlashCommandAdapter {
         Guild guild = Objects.requireNonNull(event.getGuild());
         Member bot = guild.getSelfMember();
 
-        if (!handleChecks(bot, author, targetOption.getAsMember(), guild, event)) {
+        if (!handleChecks(bot, author, targetOption.getAsMember(), event)) {
             return;
         }
 
@@ -130,7 +119,7 @@ public final class AuditCommand extends SlashCommandAdapter {
 
     @SuppressWarnings("BooleanMethodNameMustStartWithQuestion")
     private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
-            @Nullable Member target, @NotNull Guild guild, @NotNull IReplyCallback event) {
+            @Nullable Member target, @NotNull IReplyCallback event) {
         // Member doesn't exist if attempting to audit a user who is not part of the guild.
         if (target != null && !ModerationUtils.handleCanInteractWithTarget(ACTION_VERB, bot, author,
                 target, event)) {
@@ -144,21 +133,9 @@ public final class AuditCommand extends SlashCommandAdapter {
         List<ActionRecord> actions =
                 actionsStore.getActionsByTargetAscending(guild.getIdLong(), user.getIdLong());
 
-        MessageEmbed summary = createSummaryMessage(user, actions);
-        if (actions.isEmpty()) {
-            event.replyEmbeds(summary).queue();
-            return;
-        }
+        EmbedBuilder audit = createSummaryEmbed(user, actions);
+        actions.forEach(action -> audit.addField(actionToField(action, event.getJDA())));
 
-        // Computing messages for actual actions is done deferred and might require asking the
-        // Discord API
-        event.deferReply().queue();
-        JDA jda = event.getJDA();
-
-        RestAction<List<MessageEmbed>> messagesTask = RestAction
-            .allOf(actions.stream().map(action -> actionToMessage(action, jda)).toList());
-        messagesTask.map(messages -> prependElement(summary, messages))
-            .flatMap(messages -> event.getHook().sendMessageEmbeds(messages))
-            .queue();
+        event.replyEmbeds(audit.build()).queue();
     }
 }

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -125,12 +125,7 @@ public final class AuditCommand extends SlashCommandAdapter {
             return;
         }
 
-        event.reply(auditMessage(
-                guild.getIdLong(),
-                target.getIdLong(),
-                1,
-                event.getJDA()
-        )).queue();
+        event.reply(auditMessage(guild.getIdLong(), target.getIdLong(), 1, event.getJDA())).queue();
     }
 
     @SuppressWarnings("BooleanMethodNameMustStartWithQuestion")
@@ -159,24 +154,19 @@ public final class AuditCommand extends SlashCommandAdapter {
         int totalPage = groupedActions.size();
 
         EmbedBuilder audit = createSummaryEmbed(jda.retrieveUserById(targetId).complete(), actions)
-                .setFooter("Page: " + pageNo + "/" + totalPage);
-        groupedActions.get(pageNo - 1).forEach(action -> audit.addField(actionToField(action, jda)));
+            .setFooter("Page: " + pageNo + "/" + totalPage);
+        groupedActions.get(pageNo - 1)
+            .forEach(action -> audit.addField(actionToField(action, jda)));
 
         return new MessageBuilder(audit.build())
-                .setActionRows(ActionRow.of(
-                        previousButton(guildId, targetId, pageNo),
-                        nextButton(guildId, targetId, pageNo, totalPage)
-                ))
-                .build();
+            .setActionRows(ActionRow.of(previousButton(guildId, targetId, pageNo),
+                    nextButton(guildId, targetId, pageNo, totalPage)))
+            .build();
     }
 
     private Button previousButton(long guildId, long targetId, int pageNo) {
-        Button previousButton = Button.primary(generateComponentId(
-                String.valueOf(guildId),
-                String.valueOf(targetId),
-                String.valueOf(pageNo),
-                "-1"
-        ), "⬅");
+        Button previousButton = Button.primary(generateComponentId(String.valueOf(guildId),
+                String.valueOf(targetId), String.valueOf(pageNo), "-1"), "⬅");
 
         if (pageNo == 1) {
             previousButton = previousButton.asDisabled();
@@ -186,12 +176,8 @@ public final class AuditCommand extends SlashCommandAdapter {
     }
 
     private Button nextButton(long guildId, long targetId, int pageNo, int totalPage) {
-        Button nextButton = Button.primary(generateComponentId(
-                String.valueOf(guildId),
-                String.valueOf(targetId),
-                String.valueOf(pageNo),
-                "1"
-        ), "➡");
+        Button nextButton = Button.primary(generateComponentId(String.valueOf(guildId),
+                String.valueOf(targetId), String.valueOf(pageNo), "1"), "➡");
 
         if (pageNo == totalPage) {
             nextButton = nextButton.asDisabled();
@@ -202,11 +188,9 @@ public final class AuditCommand extends SlashCommandAdapter {
 
     @Override
     public void onButtonClick(@NotNull ButtonInteractionEvent event, @NotNull List<String> args) {
-        event.editMessage(auditMessage(
-                Long.parseLong(args.get(0)),
-                Long.parseLong(args.get(1)),
-                Integer.parseInt(args.get(2)) + Integer.parseInt(args.get(3)),
-                event.getJDA()
-        )).queue();
+        event
+            .editMessage(auditMessage(Long.parseLong(args.get(0)), Long.parseLong(args.get(1)),
+                    Integer.parseInt(args.get(2)) + Integer.parseInt(args.get(3)), event.getJDA()))
+            .queue();
     }
 }

--- a/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
+++ b/application/src/main/java/org/togetherjava/tjbot/commands/moderation/AuditCommand.java
@@ -152,12 +152,16 @@ public final class AuditCommand extends SlashCommandAdapter {
 
         int totalPage = groupedActions.size();
 
-        EmbedBuilder audit = createSummaryEmbed(jda.retrieveUserById(targetId).complete(), actions)
-            .setFooter("Page: " + pageNo + "/" + totalPage);
+        EmbedBuilder audit = createSummaryEmbed(jda.retrieveUserById(targetId).complete(), actions);
+
+        if (groupedActions.size() == 0) {
+            return new MessageBuilder(audit.build()).build();
+        }
+
         groupedActions.get(pageNo - 1)
             .forEach(action -> audit.addField(actionToField(action, jda)));
 
-        return new MessageBuilder(audit.build())
+        return new MessageBuilder(audit.setFooter("Page: " + pageNo + "/" + totalPage).build())
             .setActionRows(ActionRow.of(previousButton(guildId, targetId, pageNo),
                     nextButton(guildId, targetId, pageNo, totalPage)))
             .build();

